qa/tasks: be python3 compatible

tchaikov opened the pull request:
Signed-off-by: Kefu Chai <kchai@redhat.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

kshtsk commented on qa/tasks/cephfs/mount.py:
```diff
@@ -240,7 +240,7 @@ def _verify(self, proc, retval=None, errmsg=None):
 
         if errmsg:
-            stderr = proc.stderr.getvalue().lower()
+            stderr = proc.stderr.getvalue().decode().lower()
```
why not `six.ensure_str` here?

tchaikov replied:
> ah yeah, it is a unicode, still needs to use `six.ensure_str` here

could you be more specific? what is a unicode. and why do we still need to use `six.ensure_str` here?

tchaikov commented:
Le dim. 29 mars 2020 à 02:34, kshtsk <notifications@github.com> a écrit :

> *@kshtsk* commented on this pull request.
> ------------------------------
>
> In qa/tasks/cephfs/mount.py
> <https://github.com/ceph/ceph/pull/34264#discussion_r399694373>:
>
> > @@ -240,7 +240,7 @@ def _verify(self, proc, retval=None, errmsg=None):
>              assert proc.returncode == retval, msg
>
>          if errmsg:
> -            stderr = proc.stderr.getvalue().lower()
> +            stderr = proc.stderr.getvalue().decode().lower()
>
> why not six.ensure_str here?
>
I plan to drop six after migration to python3

—
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/34264#pullrequestreview-383337788>,
>
-- 
Regards
Kefu Chai

kshtsk commented on qa/tasks/cephfs/mount.py:
```diff
@@ -780,7 +780,7 @@ def getfattr(self, path, attr):
             p.wait()
         except CommandFailedError as e:
-            if e.exitstatus == 1 and "No such attribute" in p.stderr.getvalue():
+            if e.exitstatus == 1 and b"No such attribute" in p.stderr.getvalue():
```
This method:
```
    def getfattr(self, path, attr):
        """
        Wrap getfattr: return the values of a named xattr on one file, or
        None if the attribute is not found.
        :return: a string
```
But I see this:
```
        return p.stdout.getvalue()
```
Have we havde decode stdout here since we're sure the stdout is io.Bytes?
Also I can see in the code:
```
return None
```
Shouldn't we change description like:
```
        :return: a string, or None if attribute missing 
```

tchaikov replied:
i'd say this is not in the scope of this change, as the commit message does not claim to fix this issue. do you mind sending a PR to address this?

kshtsk commented:
How do you test this?

tchaikov commented:
> How do you test this?

i ran cephfs suite with this change. see #34171.

tchaikov commented:
@kshtsk @toabctl could you help approve it if it's good to go? as #34171 backport this commit, and is pending on this change. i'd love to merge nautilus change sooner so i can start working on octopus.

kshtsk commented:
@tchaikov trying to run a test agains in py3 environment, this is what I have:
```
OSError: /home/runner/src/github.com_ceph_ceph_PR#34264/qa/suites/cephfs does not exist (abs /home/runner/src/github.com_ceph_ceph_PR#34264/qa/suites/cephfs)
```
Which exact suite have you been running?

tchaikov commented:
> @tchaikov trying to run a test agains in py3 environment, this is what I have:
> 
> ```
> OSError: /home/runner/src/github.com_ceph_ceph_PR#34264/qa/suites/cephfs does not exist (abs /home/runner/src/github.com_ceph_ceph_PR#34264/qa/suites/cephfs)
> ```
> 
> Which exact suite have you been running?

fs. see http://pulpito.ceph.com/kchai-2020-03-28_05:57:39-fs-nautilus-kefu2-testing-2020-03-27-2234-distro-basic-smithi/

kshtsk commented:
Have many errors like this:
```
2020-03-30T17:37:06.675 INFO:tasks.cephfs_test_runner:test_flush (tasks.cephfs.test_flush.TestFlush) ... ERROR
2020-03-30T17:37:06.676 INFO:tasks.cephfs_test_runner:
2020-03-30T17:37:06.677 INFO:tasks.cephfs_test_runner:======================================================================
2020-03-30T17:37:06.678 INFO:tasks.cephfs_test_runner:ERROR: test_flush (tasks.cephfs.test_flush.TestFlush)
2020-03-30T17:37:06.679 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-03-30T17:37:06.680 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-03-30T17:37:06.681 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs/cephfs_test_case.py", line 138, in setUp
2020-03-30T17:37:06.681 INFO:tasks.cephfs_test_runner:    'osd', 'allow rw pool={0}'.format(self.fs.get_data_pool_name()))
2020-03-30T17:37:06.682 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs/filesystem.py", line 694, in get_data_pool_name
2020-03-30T17:37:06.683 INFO:tasks.cephfs_test_runner:    return self.data_pools.values()[0]
2020-03-30T17:37:06.684 INFO:tasks.cephfs_test_runner:TypeError: 'dict_values' object does not support indexing
2020-03-30T17:37:06.684 INFO:tasks.cephfs_test_runner:
2020-03-30T17:37:06.685 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-03-30T17:37:06.686 INFO:tasks.cephfs_test_runner:Ran 1 test in 20.896s
2020-03-30T17:37:06.686 INFO:tasks.cephfs_test_runner:
2020-03-30T17:37:06.687 INFO:tasks.cephfs_test_runner:FAILED (errors=1)
2020-03-30T17:37:06.688 INFO:tasks.cephfs_test_runner:
2020-03-30T17:37:06.688 INFO:tasks.cephfs_test_runner:======================================================================
2020-03-30T17:37:06.689 INFO:tasks.cephfs_test_runner:ERROR: test_flush (tasks.cephfs.test_flush.TestFlush)
2020-03-30T17:37:06.689 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-03-30T17:37:06.689 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-03-30T17:37:06.690 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs/cephfs_test_case.py", line 138, in setUp
2020-03-30T17:37:06.690 INFO:tasks.cephfs_test_runner:    'osd', 'allow rw pool={0}'.format(self.fs.get_data_pool_name()))
2020-03-30T17:37:06.691 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs/filesystem.py", line 694, in get_data_pool_name
2020-03-30T17:37:06.691 INFO:tasks.cephfs_test_runner:    return self.data_pools.values()[0]
2020-03-30T17:37:06.692 INFO:tasks.cephfs_test_runner:TypeError: 'dict_values' object does not support indexing
2020-03-30T17:37:06.692 INFO:tasks.cephfs_test_runner:
2020-03-30T17:37:06.693 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/worker/src/teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
    manager.__enter__()
  File "/usr/lib64/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs_test_runner.py", line 211, in task
    raise RuntimeError("Test failure: {0}".format(", ".join(bad_tests)))
RuntimeError: Test failure: test_flush (tasks.cephfs.test_flush.TestFlush)
```
http://51.68.86.123:8081/runner-2020-03-30_17:16:11-fs:basic_functional-master-distro-basic-ovh/

tchaikov commented:
@kshtsk the latest change should fix this issue.

tchaikov commented:
```
2020-03-31T11:32:45.745 INFO:tasks.cephfs_test_runner:ERROR: test_fragmented_injection (tasks.cephfs.test_data_scan.TestDataScan)
2020-03-31T11:32:45.746 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-03-31T11:32:45.747 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-03-31T11:32:45.747 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs/test_data_scan.py", line 474, in test_fragmented_injection
2020-03-31T11:32:45.748 INFO:tasks.cephfs_test_runner:    files = self.mount_a.run_shell(["ls", "subdir/"]).stdout.getvalue().strip().split("\n")
2020-03-31T11:32:45.749 INFO:tasks.cephfs_test_runner:TypeError: a bytes-like object is required, not 'str'
2020-03-31T11:32:45.750 INFO:tasks.cephfs_test_runner:
2020-03-31T11:32:45.750 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-03-31T11:32:45.751 INFO:tasks.cephfs_test_runner:Ran 1 test in 137.612s
2020-03-31T11:32:45.751 INFO:tasks.cephfs_test_runner:
```
should fix the callers of `run_shell()` and `_run_python()`

kshtsk commented:
```2020-04-01T16:32:53.714 INFO:tasks.cephfs_test_runner:test_replay_create (tasks.cephfs.test_cap_flush.TestCapFlush) ... FAIL
2020-04-01T16:32:53.715 INFO:tasks.cephfs_test_runner:
2020-04-01T16:32:53.717 INFO:tasks.cephfs_test_runner:======================================================================
2020-04-01T16:32:53.718 INFO:tasks.cephfs_test_runner:FAIL: test_replay_create (tasks.cephfs.test_cap_flush.TestCapFlush)
2020-04-01T16:32:53.719 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-04-01T16:32:53.720 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-04-01T16:32:53.721 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/test_cap_flush.py", line 63, in test_replay_create
2020-04-01T16:32:53.722 INFO:tasks.cephfs_test_runner:    self.assertEqual(mode, "600")
2020-04-01T16:32:53.723 INFO:tasks.cephfs_test_runner:AssertionError: b'640' != '600'
2020-04-01T16:32:53.723 INFO:tasks.cephfs_test_runner:
2020-04-01T16:32:53.724 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-04-01T16:32:53.725 INFO:tasks.cephfs_test_runner:Ran 1 test in 126.678s
2020-04-01T16:32:53.726 INFO:tasks.cephfs_test_runner:
2020-04-01T16:32:53.727 INFO:tasks.cephfs_test_runner:FAILED (failures=1)
2020-04-01T16:32:53.727 INFO:tasks.cephfs_test_runner:
2020-04-01T16:32:53.728 INFO:tasks.cephfs_test_runner:======================================================================
2020-04-01T16:32:53.728 INFO:tasks.cephfs_test_runner:FAIL: test_replay_create (tasks.cephfs.test_cap_flush.TestCapFlush)
2020-04-01T16:32:53.728 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-04-01T16:32:53.729 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-04-01T16:32:53.729 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/test_cap_flush.py", line 63, in test_replay_create
2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:    self.assertEqual(mode, "600")
2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:AssertionError: b'640' != '600'
2020-04-01T16:32:53.731 INFO:tasks.cephfs_test_runner:
2020-04-01T16:32:53.731 ERROR:teuthology.run_tasks:Saw exception from tasks.
Traceback (most recent call last):
  File "/home/worker/src/teuthology_master/teuthology/run_tasks.py", line 89, in run_tasks
    manager.__enter__()
  File "/usr/lib64/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs_test_runner.py", line 211, in task
    raise RuntimeError("Test failure: {0}".format(", ".join(bad_tests)))
RuntimeError: Test failure: test_replay_create (tasks.cephfs.test_cap_flush.TestCapFlush)```

@tchaikov Can see this ^
http://51.68.86.123/teuthology/runner-2020-04-01_16:07:16-fs:basic_functional-master-distro-basic-ovh/87/teuthology.log

kshtsk commented:
another failure:
```
ERROR
2020-04-01T17:22:09.273 INFO:tasks.cephfs_test_runner:
2020-04-01T17:22:09.274 INFO:tasks.cephfs_test_runner:======================================================================
2020-04-01T17:22:09.275 INFO:tasks.cephfs_test_runner:ERROR: test_15303 (tasks.cephfs.test_volume_client.TestVolumeClient)
2020-04-01T17:22:09.276 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-04-01T17:22:09.276 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-04-01T17:22:09.277 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/test_volume_client.py", line 404, in test_15303
2020-04-01T17:22:09.278 INFO:tasks.cephfs_test_runner:    self.assertListEqual(self.mount_a.ls("parent1/mydir"),
2020-04-01T17:22:09.279 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/mount.py", line 755, in ls
2020-04-01T17:22:09.280 INFO:tasks.cephfs_test_runner:    return ls_text.split("\n")
2020-04-01T17:22:09.280 INFO:tasks.cephfs_test_runner:TypeError: a bytes-like object is required, not 'str'
2020-04-01T17:22:09.281 INFO:tasks.cephfs_test_runner:
2020-04-01T17:22:09.281 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-04-01T17:22:09.282 INFO:tasks.cephfs_test_runner:Ran 1 test in 148.383s
2020-04-01T17:22:09.282 INFO:tasks.cephfs_test_runner:
2020-04-01T17:22:09.283 INFO:tasks.cephfs_test_runner:FAILED (errors=1)
2020-04-01T17:22:09.284 INFO:tasks.cephfs_test_runner:
2020-04-01T17:22:09.284 INFO:tasks.cephfs_test_runner:======================================================================
2020-04-01T17:22:09.285 INFO:tasks.cephfs_test_runner:ERROR: test_15303 (tasks.cephfs.test_volume_client.TestVolumeClient)
2020-04-01T17:22:09.286 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2020-04-01T17:22:09.286 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2020-04-01T17:22:09.287 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/test_volume_client.py", line 404, in test_15303
2020-04-01T17:22:09.288 INFO:tasks.cephfs_test_runner:    self.assertListEqual(self.mount_a.ls("parent1/mydir"),
2020-04-01T17:22:09.288 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/mount.py", line 755, in ls
2020-04-01T17:22:09.288 INFO:tasks.cephfs_test_runner:    return ls_text.split("\n")
2020-04-01T17:22:09.289 INFO:tasks.cephfs_test_runner:TypeError: a bytes-like object is required, not 'str'
2020-04-01T17:22:09.289 INFO:tasks.cephfs_test_runner:
```

tchaikov commented:
> another failure:
> 
> ```
> ERROR
> 2020-04-01T17:22:09.273 INFO:tasks.cephfs_test_runner:
> 2020-04-01T17:22:09.274 INFO:tasks.cephfs_test_runner:======================================================================
> 2020-04-01T17:22:09.275 INFO:tasks.cephfs_test_runner:ERROR: test_15303 (tasks.cephfs.test_volume_client.TestVolumeClient)
> 2020-04-01T17:22:09.276 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-04-01T17:22:09.276 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2020-04-01T17:22:09.277 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/test_volume_client.py", line 404, in test_15303
> 2020-04-01T17:22:09.278 INFO:tasks.cephfs_test_runner:    self.assertListEqual(self.mount_a.ls("parent1/mydir"),
> 2020-04-01T17:22:09.279 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/mount.py", line 755, in ls
> 2020-04-01T17:22:09.280 INFO:tasks.cephfs_test_runner:    return ls_text.split("\n")
> 2020-04-01T17:22:09.280 INFO:tasks.cephfs_test_runner:TypeError: a bytes-like object is required, not 'str'
> 2020-04-01T17:22:09.281 INFO:tasks.cephfs_test_runner:
> 2020-04-01T17:22:09.281 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-04-01T17:22:09.282 INFO:tasks.cephfs_test_runner:Ran 1 test in 148.383s
> 2020-04-01T17:22:09.282 INFO:tasks.cephfs_test_runner:
> 2020-04-01T17:22:09.283 INFO:tasks.cephfs_test_runner:FAILED (errors=1)
> 2020-04-01T17:22:09.284 INFO:tasks.cephfs_test_runner:
> 2020-04-01T17:22:09.284 INFO:tasks.cephfs_test_runner:======================================================================
> 2020-04-01T17:22:09.285 INFO:tasks.cephfs_test_runner:ERROR: test_15303 (tasks.cephfs.test_volume_client.TestVolumeClient)
> 2020-04-01T17:22:09.286 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-04-01T17:22:09.286 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2020-04-01T17:22:09.287 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/test_volume_client.py", line 404, in test_15303
> 2020-04-01T17:22:09.288 INFO:tasks.cephfs_test_runner:    self.assertListEqual(self.mount_a.ls("parent1/mydir"),
> 2020-04-01T17:22:09.288 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_tchaikov_ceph_wip-qa-py3/qa/tasks/cephfs/mount.py", line 755, in ls
> 2020-04-01T17:22:09.288 INFO:tasks.cephfs_test_runner:    return ls_text.split("\n")
> 2020-04-01T17:22:09.289 INFO:tasks.cephfs_test_runner:TypeError: a bytes-like object is required, not 'str'
> 2020-04-01T17:22:09.289 INFO:tasks.cephfs_test_runner:
> ```

this is a todo. see https://github.com/ceph/ceph/pull/34264#issuecomment-606592305

sebastian-philipp commented on qa/tasks/cephfs/mount.py:
```diff
@@ -240,7 +240,7 @@ def _verify(self, proc, retval=None, errmsg=None):
 
         if errmsg:
-            stderr = proc.stderr.getvalue().lower()
+            stderr = proc.stderr.getvalue().decode().lower()
```
```suggestion
            stderr = proc.stderr.getvalue().decode('utf-8').lower()
```

`decode` in python2 might default to something completely wrong.

kshtsk replied:
ah yeah, it is a unicode, still needs to use `six.ensure_str` here

tchaikov replied:
this part is still in flux. see #34386. i am inclined to hold back until these tests are settled.

tchaikov replied:
> ah yeah, it is a unicode, still needs to use `six.ensure_str` here

@kshtsk could you be more specific? what is a unicode. and why do we still need to use `six.ensure_str` here?

kshtsk replied:
@tchaikov I meant this:
```
>>> a=b'a'
>>> a.decode()
u'a'
```

kshtsk replied:
@tchaikov it is python2 specific class
```
Help on class unicode in module __builtin__:

class unicode(basestring)
 |  unicode(object='') -> unicode object
 |  unicode(string[, encoding[, errors]]) -> unicode object
```

kshtsk replied:
so *decode()* returns a _unicode_ which is not a standard _str_ which is actually _bytes_, latest python2 can transparently convert `str` to `unicode` in expressions where the latter occurs, which can make an unexpected results where it does matter.

tchaikov replied:
thanks @kshtsk and @sebastian-philipp . i am using `StringIO` to address this issue.

sebastian-philipp commented on qa/tasks/cephfs/test_damage.py:
```diff
@@ -330,7 +330,7 @@ def __hash__(self):
                     proc.wait()
                 except CommandFailedError:
-                    stderr = proc.stderr.getvalue()
+                    stderr = proc.stderr.getvalue().decode()
```
same

tchaikov replied:
i will continue working on the `run_shell()` related changes after ##34386 is merged.

tchaikov replied:
ditto, use `StringIO` now.

kshtsk commented:
> ```
> 020-03-31T11:32:45.746 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-03-31T11:32:45.747 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2020-03-31T11:32:45.747 INFO:tasks.cephfs_test_runner:  File "/home/worker/src/github.com_ceph_ceph_PR#34264/qa/tasks/cephfs/test_data_scan.py", line 474, in test_fragmented_injection
> 2020-03-31T11:32:45.748 INFO:tasks.cephfs_test_runner:    files = self.mount_a.run_shell(["ls", "subdir/"]).stdout.getvalue().strip().split("\n")
> 2020-03-31T11:32:45.749 INFO:tasks.cephfs_test_runner:TypeError: a bytes-like object is required, not 'str'
> 2020-03-31T11:32:45.750 INFO:tasks.cephfs_test_runner:
> 2020-03-31T11:32:45.750 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
> 2020-03-31T11:32:45.751 INFO:tasks.cephfs_test_runner:Ran 1 test in 137.612s
> 2020-03-31T11:32:45.751 INFO:tasks.cephfs_test_runner:
> ```
> 
> 
> 
>     *

I wonder why do we still need run_shell? I think it is enough in many cases just `remote.sh`, it has all you need.

tchaikov commented:
@sebastian-philipp @kshtsk @toabctl i am moving part of this changeset into #34389 in hope to get some of the change merged sooner. could you help review it? see #34389

tchaikov commented:
since #34386 has been merged. i am closing #34389, will continue following up in this PR.

kshtsk commented on qa/tasks/cephfs/test_openfiletable.py:
```diff
@@ -31,7 +32,7 @@ def test_max_items_per_obj(self):
         # Hold the file open
         file_count = 8
-        for i in xrange(0, file_count):
+        for i in range(0, file_count):
```
file_count is 8, there is no need to use `xrange` and `six.moves.range` correspondingly

tchaikov replied:
ack. fixed.

kshtsk commented on src/test/rgw/rgw_multi/tests_es.py:
```diff
@@ -77,7 +78,7 @@ def init_env(create_obj, num_keys = 5, buckets_per_zone = 1, bucket_init_cb = No
     # don't wait for meta sync just yet
     for zone, bucket in zone_bucket:
-        for count in xrange(0, num_keys):
+        for count in range(num_keys):
```
default num_keys is 5, not sure it ever more then 1000 or so... no need to use xrange here.

tchaikov replied:
@kshtsk i don't want to take the risk. if we are not sure, why would we want to do this?

tchaikov replied:
please also see my reply at https://github.com/ceph/ceph/pull/34389#discussion_r403484551 . again, what i am aiming is the ultimate goal. personally, i don't really want take the risk or even the time to figure out the upper bound of `num_keys`.

kshtsk replied:
ok

kshtsk commented on src/test/rgw/rgw_multi/tests_es.py:
```diff
@@ -226,16 +227,16 @@ def create_obj(k, i):
                 do_check_mdsearch(target_conn.conn, bucket, src_keys , 'x-amz-meta-foo-date > ' + sval, lambda k: date_from_str(k.get_metadata('foo-date')) > val)
 
             # 'or' query
-            for i in xrange(len(src_keys) / 2):
```
Not sure src_keys has a whatever big value... so no need in xrange and six.move.range here

tchaikov replied:
ditto.

tchaikov commented:
> 2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:AssertionError: b'640' != '600'

@kshtsk i am not sure about this. will rerun the test without the py3 changes to understand it better.

tchaikov commented:
- http://pulpito.ceph.com/kchai-2020-04-03_15:17:08-fs-master-distro-basic-smithi/

kshtsk commented:
> > 2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:AssertionError: b'640' != '600'
> 
> @kshtsk i am not sure about this. will rerun the test without the py3 changes to understand it better.

@tchaikov I guess you need to change this:
```
self.assertEqual(mode, "600")
```
on this:
```
self.assertEqual(mode, b'600')
```
Or use `six.ensure_str`.

tchaikov commented:
> > > 2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:AssertionError: b'640' != '600'
> > 
> > 
> > @kshtsk i am not sure about this. will rerun the test without the py3 changes to understand it better.
> 
> @tchaikov I guess you need to change this:
> 
> ```
> self.assertEqual(mode, "600")
> ```
> 
> into:
> 
> ```
> self.assertEqual(mode, b'600')
> ```
> 
> Or use `six.ensure_str`.

what i wanted to figure out is why we have "640" but not "600".

```
2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:    self.assertEqual(mode, "600")
2020-04-01T16:32:53.730 INFO:tasks.cephfs_test_runner:AssertionError: b'640' != '600'
```

tchaikov commented:
http://pulpito.ceph.com/kchai-2020-04-05_11:22:54-fs:basic_functional-master-distro-basic-smithi/

kshtsk commented:
> what i wanted to figure out is why we have "640" but not "600".
Ah... indeed :-D

kshtsk commented on qa/tasks/dnsmasq.py:
```diff
@@ -83,7 +83,7 @@ def setup_dnsmasq(remote, testdir, cnames):
     remote.run(args=['sudo', 'systemctl', 'restart', 'dnsmasq'])
     # verify dns name is set
-    remote.run(args=['ping', '-c', '4', cnames.keys()[0]])
+    remote.run(args=['ping', '-c', '4', next(iter(cnames.keys()))])
```
(nit) Hm, wonder why the code is checking only first address, can we fix it here: to cycle throw all cnames?

tchaikov replied:
this is not in the scope of this PR, i'd prefer deferring it after we get this change merged and having it in a following-up.

tchaikov commented:
- ceph=master [0], teuthology=mater, suite=master : http://pulpito.ceph.com/kchai-2020-04-05_17:12:09-fs-master-distro-basic-smithi/
- ceph=master, teuthology=wip-py3[1], suite=wip-qa-py3: http://pulpito.ceph.com/kchai-2020-04-06_01:38:10-fs-master-distro-basic-smithi/

---
[0] de4ca87f61ccc5d1830fdfdf3659b617a7573fc2
[1] https://github.com/ceph/teuthology/pull/1439 + https://github.com/ceph/teuthology/pull/1422

tchaikov commented:
jenkins test make check

kshtsk commented:
@tchaikov do you want me to run some extra tests for this PR?

kshtsk commented on src/test/rgw/rgw_multi/zone_ps.py:
```diff
@@ -72,7 +72,7 @@ def make_request(conn, method, resource, parameters=None, sign_parameters=False,
     url_params = ''
     if parameters is not None:
-        url_params = urllib.urlencode(parameters)
+        url_params = urlparse.urlencode(parameters)
```
Just a reminder that one can use this:
```
from teuthology.util.compat import urlencode
.
.
.
url_params = urlencode(parameters)
```

kshtsk commented on qa/tasks/cephadm.py:
```diff
@@ -391,7 +391,7 @@ def ceph_bootstrap(ctx, config):
             remote=bootstrap_remote,
             path='{}/{}.pub'.format(testdir, cluster_name)
-        ).strip()
+        ).decode('ascii').strip()
```
Not sure I understand why we need to decode this? From my understanding the keys should be always binary.

kshtsk replied:
Ah... it is literally cat it later...

kshtsk commented:
we will need to backport it to octopus first I guess

tchaikov commented:
Le lun. 6 avr. 2020 à 23:38, kshtsk <notifications@github.com> a écrit :

> @tchaikov <https://github.com/tchaikov> do you want me to run some extra
> tests for this PR?
>
Thanks. That’d be nice, but I’d need an approval so I can merge this big
and fat PR sooner.

—
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceph/ceph/pull/34264#issuecomment-609870206>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAONP45PTFIWUQKIVYYGQTRLHZORANCNFSM4LVRE6GQ>
> .
>
-- 
Regards
Kefu Chai

tchaikov commented:
rados suite
- http://pulpito.ceph.com/kchai-2020-04-07_09:41:02-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-07_15:13:18-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-08_01:39:26-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/

tchaikov commented:
rados suite
- http://pulpito.ceph.com/kchai-2020-04-07_09:41:02-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-07_15:13:18-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-08_01:39:26-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-08_03:18:15-rados-wip-kefu-testing-2020-04-07-1603-distro-basic-smithi/

tchaikov commented:
rgw suite

- http://pulpito.ceph.com/kchai-2020-04-06_02:54:33-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-06_05:28:37-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-06_07:21:00-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-06_11:00:12-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-06_13:54:03-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-06_15:22:16-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-06_16:53:23-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-07_04:25:12-rgw-master-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-07_04:25:12-rgw-master-distro-basic-smithi/

tchaikov commented:
rerunning tests with all fixes

- http://pulpito.ceph.com/kchai-2020-04-08_07:25:49-rados-wip-kefu-testing-2020-04-08-1234-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-08_07:22:38-rgw-wip-kefu-testing-2020-04-08-1234-distro-basic-smithi/

toabctl commented on qa/tasks/ceph_manager.py:
```diff
@@ -1965,7 +1965,7 @@ def get_pool(self):
         with self.lock:
             if self.pools:
-                return random.choice(self.pools.keys())
+                return random.sample(self.pools.keys(), 1)[0]
```
shouldn't you use `iter` here to access the first element?

tchaikov replied:
no need to do so, both Python2 and Python3 return a list, so we are safe here.

tchaikov replied:
```
>>> print(random.sample.__doc__)
Chooses k unique random elements from a population sequence.

        Returns a new list containing elements from the population while
        leaving the original population unchanged.  The resulting list is
        in selection order so that all sub-slices will also be valid random
        samples.  This allows raffle winners (the sample) to be partitioned
        into grand prize and second place winners (the subslices).
```

toabctl commented on qa/tasks/cephfs/test_dump_tree.py:
```diff
@@ -39,7 +39,7 @@ def test_random(self):
         self.populate()
         inos = self.get_paths_to_ino()
-        target = random.choice(inos.keys())
+        target = random.sample(inos.keys(), 1)[0]
```
same here (shouldn't you use `iter` here to access the first element?)

tchaikov replied:
ditto.

tchaikov commented on qa/tasks/openssl_keys.py:
```diff
@@ -64,8 +64,9 @@ def setup(self):
-        for name, config in self.config.items():
+        # make sure self-signed certs get added first, they don't have 'ca' field
+        configs = sorted(self.config.items(), key=lambda x: 'ca' in x[1])
+        for name, config in configs:
```
@cbodley could you help review this part?

tchaikov commented:
rerunning tests with all fixes

- http://pulpito.ceph.com/kchai-2020-04-08_07:25:49-rados-wip-kefu-testing-2020-04-08-1234-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2020-04-08_07:22:38-rgw-wip-kefu-testing-2020-04-08-1234-distro-basic-smithi/
  - http://pulpito.ceph.com/kchai-2020-04-08_13:07:30-rgw-master-distro-basic-smithi

tchaikov commented:
@jdurgin thanks for reviewing this change!

jdurgin commented:
np, that was a fast merge! can't say if there's more to go or not, but everything in this PR looked correct

tchaikov commented:
i will keep an eye on the results from master on pulpito. and switch teuthology to py3 ASAP.
